compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -368,11 +391,103 @@ impl<'tcx> SizeSkeleton<'tcx> {
             (SizeSkeleton::Pointer { tail: a, .. }, SizeSkeleton::Pointer { tail: b, .. }) => {
                 a == b
             }
+            // constants are always pre-normalized into a canonical form so this
+            // only needs to check if their pointers are identical.
+            (SizeSkeleton::Generic(a), SizeSkeleton::Generic(b)) => a == b,
             _ => false,
         }
     }
 }
 
+fn combine_sorted_expr<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    a: ty::Const<'tcx>,
+    b: ty::Const<'tcx>,
+) -> ty::Const<'tcx> {
+    use crate::mir::BinOp::Mul;
+    use ty::ConstKind::Expr;
+    use ty::Expr::Binop;
+
+    // fast-path
+    match (a.kind(), b.kind()) {
+        (Expr(Binop(Mul, ..)), _) | (_, Expr(Binop(Mul, ..))) => {}
+        _ => {
+            let (l, r) = if a < b { (a, b) } else { (b, a) };
+            return tcx.mk_const(Expr(Binop(Mul, l, r)), tcx.types.usize);
+        }
+    }
+
+    let mut buf = vec![];
+    let mut a = sort_mul(tcx, param_env, a);
+    let mut b = sort_mul(tcx, param_env, b);
+    while let (Expr(Binop(Mul, al, ar)), Expr(Binop(Mul, bl, br))) = (a.kind(), b.kind()) {
+        if al < bl {
+            buf.push(al);
+            a = ar;
+        } else {
+            buf.push(bl);
+            b = br;
+        }
+    }
+
+    while let Expr(Binop(Mul, al, ar)) = a.kind() {
+        if al < b {
+            buf.push(al);
+        } else {
+            buf.push(std::mem::replace(&mut b, al));
+        }
+        a = ar;
+    }
+    while let Expr(Binop(Mul, bl, br)) = b.kind() {
+        if bl < a {
+            buf.push(bl);
+        } else {
+            buf.push(std::mem::replace(&mut a, bl));
+        }
+        b = br;
+    }
+
+    if a < b {
+        buf.push(a);
+        buf.push(b);
+    } else {
+        buf.push(b);
+        buf.push(a);
+    }
+
+    // buf is now sorted minimum -> maximum, with evaluatable at the end.
+    // compress all multiplied constants into a single value.
+    let mut const_mul = 1;
+    while let Some(v) = buf.last().and_then(|c| c.try_eval_usize(tcx, param_env)) {
+        const_mul *= v;
```

Comment:
This should be a `checked_mul` or go through `Size`, you don't want to let through overflows.